### ISSUE TITLE
softmax, categorical, multinomial with K-1 parameters for identifiability

### ISSUE THREAD

**Issue by alashworth** [alashworth/stan-monorepo-stage1#3]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [bob-carpenter](https://github.com/bob-carpenter)**
_Tuesday Oct 08, 2013 at 23:03 GMT_
_Originally opened as https://github.com/stan-dev/stan/issues/266_

----

Add new versions of all these functions for the (K-1)-vector parameterization of softmax.

```
softmax_dofm1(vector x)  =def=  softmax(c(x,0) )
```

where `c(x,0)` is the R function to append 0 to the end of x.

The advantage is that the degrees of freedom matches that of the simplex target, so the model's identified without priors.

The disadvantage is that the priors have to change relative to the K-value parameterization and are no longer symmetric.
- [ ] Any suggestions for a better name than `dofm`?  I suppose we could say `identified`.  
- [ ] Any preference for `c(x,0)` or `c(0,x)`, i.e., should the 0 value be the first or the last element?

**Comment by alashworth** [comment]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bob-carpenter](https://github.com/bob-carpenter)**
_Tuesday Oct 08, 2013 at 23:07 GMT_

----

The current workaround requires too much index fiddling.

```
vector[K-1] alpha_dofm1;
...
{
  vector[K] alpha;
  for (k in 1:(K-1))
    alpha[k] <- alpha_dofm1[k];
  alpha[K] <- 0;
  ... use softmax(alpha) as before --- it is a K-simplex ...
}
```

Note that `alpha` is declared in a new block to make it a local variable.

**Comment by alashworth** [comment]
<a href="https://github.com/ariddell"><img src="https://avatars2.githubusercontent.com/u/19950?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ariddell](https://github.com/ariddell)**
_Sunday Jul 20, 2014 at 15:34 GMT_

----

Should we perhaps add this workaround to the manual? Apart from the verbosity is there a reason that those interested in using the K-1 parameterization shouldn't start using this?

IMHO, anyone who can write a Stan model that uses softmax or categorical_logit can handle this without any problems.

**Comment by alashworth** [comment]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bob-carpenter](https://github.com/bob-carpenter)**
_Monday Jul 28, 2014 at 16:07 GMT_

----

This will be much easier with the equivalent of the new append operations (see pull request #790).   If `K` is the number of dimensions for the simplexes, then

```
parameters {
  matrix[N, K-1] alpha_raw;
}
model {
  matrix[N, K] alpha;
  alpha <- col_append(alpha_raw, rep_vector(0, N));  
  ...
  y[n] ~ categorical_logit(alpha[n]);
}
```

There shouldn't be any need to actually use the softmax operation directly.  Of course, users might want to see the probabilities, at which case, the generated quantities could be used (if `alpha` is a transformed parameter---otherwise it would need to be redefined).

```
generated quantities {
   simplex[K] theta[N];
    for (n in 1:N) theta[n] <- softamx(alpha[n]);
}
```

**Comment by alashworth** [comment]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bob-carpenter](https://github.com/bob-carpenter)**
_Monday Jul 28, 2014 at 16:10 GMT_

----

This would all be easier if we had
1.  an append operation that took a vector and a scalar
2.  vectorization of categorical_logit to take in an array of vectors --- we don't have the support classes for this, which is why it's not there yet, but it could be added (see how multivariate normals are done and perhaps consolidate the code)

**Comment by alashworth** [comment]
<a href="https://github.com/ariddell"><img src="https://avatars2.githubusercontent.com/u/19950?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ariddell](https://github.com/ariddell)**
_Monday Jul 28, 2014 at 16:56 GMT_

----

I am working on a multi-level varying-intercept, varying-slope categorical logit model right now. I tried the K-1 parameterization but it was painful. I take it `append_col` is slated for 2.4++?

Right now I'm inclined to try identifying the model using redundant parameters. I'd much prefer the K-1 method.

**Comment by alashworth** [comment]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bob-carpenter](https://github.com/bob-carpenter)**
_Monday Jul 28, 2014 at 18:43 GMT_

----

I think the commit went in after 2.4.  So it's
in the current dev branch.
- Bob

**Comment by alashworth** [comment]
<a href="https://github.com/randommm"><img src="https://avatars1.githubusercontent.com/u/4267674?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [randommm](https://github.com/randommm)**
_Monday Jul 28, 2014 at 19:03 GMT_

----

yep.

**Comment by alashworth** [comment]
<a href="https://github.com/ariddell"><img src="https://avatars2.githubusercontent.com/u/19950?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ariddell](https://github.com/ariddell)**
_Monday Jul 28, 2014 at 21:36 GMT_

----

Ok. I'll be a tester for the `append_col` approach. It looks good. And I certainly prefer `raw` to `dofm1`.

**Comment by alashworth** [comment]
<a href="https://github.com/bob-carpenter"><img src="https://avatars1.githubusercontent.com/u/3383807?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [bob-carpenter](https://github.com/bob-carpenter)**
_Monday Jul 28, 2014 at 21:57 GMT_

----

I don't think "raw" is descriptive enough.  I agree that "dof" is
confusing, though, and don't have a good suggestion about what to call
a new function.
